Promote exit tests to API (#324)
swiftlang/swift-testing: Sources/Testing/Testing.docc/exit-testing.md

# Exit testing

<!--
This source file is part of the Swift.org open source project

Copyright (c) 2023–2025 Apple Inc. and the Swift project authors
Licensed under Apache License v2.0 with Runtime Library Exception

See https://swift.org/LICENSE.txt for license information
See https://swift.org/CONTRIBUTORS.txt for Swift project authors
-->

@Metadata {
  @Available(Swift, introduced: 6.2)
}

Use exit tests to test functionality that may cause a test process to exit.

## Overview

> Your code may contain calls to [`precondition()`](https://developer.apple.com/documentation/swift/precondition(_:_:file:line:)),

Author: I'm having trouble building this out in a way that doesn't feel contrived ("you might write code that crashes, maybe you want to test that code yes?")

Could use some brainstorming here.

> Your code may contain calls to [`precondition()`](https://developer.apple.com/documentation/swift/precondition(_:_:file:line:)),

Reviewer: "Crashes" sounds like an unexpected end. So you wouldn't want an exit test, in that case; you'd want a regular test that expects your code to run to completion, and fails if the code crashes.

> Your code may contain calls to [`precondition()`](https://developer.apple.com/documentation/swift/precondition(_:_:file:line:)),

Author: I'm going to leave the overview as-is for the moment—let's still plan to revise it before Swift 6.2 ships.

your functions correctly catch invalid inputs.

- Note: Exit tests are available on macOS, Linux, FreeBSD, OpenBSD, and Windows.

### Create an exit test

To create an exit test, call either the ``expect(exitsWith:observing:_:sourceLocation:performing:)``
or the ``require(exitsWith:observing:_:sourceLocation:performing:)`` macro:

```swift
@Test func `Customer won't eat food unless it's delicious`() async {
  let result = await #expect(exitsWith: .failure) {
    var food = ...
    food.isDelicious = false
    Customer.current.eat(food)
  }
}
```

The closure or function reference you pass to these macros is the body of the
exit test. When an exit test is performed at runtime, the testing library starts
a new process with the same executable as the current process. The current task
is then suspended (as with `await`) and waits for the child process to
exit. The exit test's body is not called in the parent process.

Meanwhile, in the child process, the body is called directly. To ensure a clean
environment for execution, the body is not called within the context of the
original test. Instead, it is treated as if it were the `main()` function of the

> child process.

Author: They should assume nothing is inherited. The set of open file descriptors is platform-dependent but we try to close as many as we can. And, due to the nature of process spawning via POSIX API, the environment is inherited but may not exactly match that of the parent.

> child process.

Reviewer: Is there value in listing that out for developers?

If the body returns before the child process exits, it is allowed to return and

> the process exits naturally. If an error is thrown from the body, it is handled

Author: "Don't know" and "nothing".

> as if the error were thrown from `main()` and the process is forced to exit.

Author: Within the exit test, the developer can write more or less any valid Swift code. Installing uncaught exception handlers is not a Swift concept (but we don't actively block it.) Same with signal handlers—if you install your own, you're out of the realm of Swift and we make no guarantees (but it's not UB _per se_.)

`atexit()`/`std::terminate()` behave the same way they would in any other process and I don't know if they're worth calling out.

> as if the error were thrown from `main()` and the process is forced to exit.

Author: Exit handlers will still run if they are set in the child process and if the nature of the process exit would normally allow for it.

or ``require(exitsWith:observing:_:sourceLocation:performing:)``:

```swift
extension Customer {
  func eat(_ food: consuming some Food) {
    print("Let's see if I want to eat \(food)...")
    precondition(food.isDelicious, "Tasty food only!")
    precondition(food.isNutritious, "Healthy food only!")
    ...
  }
}

@Test func `Customer won't eat food unless it's delicious`() async {
  let result = await #expect(
    exitsWith: .failure,
    observing: [\.standardOutputContent]
  ) {
    var food = ...
    food.isDelicious = false
    Customer.current.eat(food)
  }
  if let result {
    #expect(result.standardOutputContent.contains(UInt8(ascii: "L")))
  }
}
```

- Note: The content of the standard output and standard error streams may
  contain any arbitrary sequence of bytes, including sequences that are not
  valid UTF-8 and cannot be decoded by [`String.init(cString:)`](https://developer.apple.com/documentation/swift/string/init(cstring:)-6kr8s).
  These streams are globally accessible within the child process, and any code
  running in an exit test may write to it including the operating system and any

>   third-party dependencies you have declared in your package.

Author: Also added "or Xcode project" here.

The testing library always sets ``ExitTest/Result/statusAtExit`` to the actual
exit status of the child process (as reported by the system) even if you do not
pass it.

> ### Constraints on exit tests

Reviewer: ```suggestion
### Write exit tests within constraints
```
…or something else that tells the developer what to do.

Author: This section doesn't tell the developer what to do. It is enumerating constraints on the code they write.

> #### State cannot be captured

Reviewer: ```suggestion
#### Don't capture state in the test body
```

Author: This isn't an instruction. It's physically impossible to do it. There is no "do".

Author: Refactored this section and moved it up.

> Exit tests cannot capture any state originating in the parent process or from

Reviewer: ```suggestion
Exit tests can't capture any state originating in the parent process or from
```

> the enclosing lexical context. For example, the following exit test will fail to

Reviewer: ```suggestion
the enclosing lexical context. For example, the following exit test fails to
```

compile because it captures a variable declared outside the exit test itself:

```swift
@Test func `Customer won't eat food unless it's nutritious`() async {
  let isNutritious = false
  await #expect(exitsWith: .failure) {
    var food = ...
    food.isNutritious = isNutritious // ❌ ERROR: trying to capture state here
    Customer.current.eat(food)
  }
}
```

> #### Exit tests cannot be nested

Reviewer: ```suggestion
#### Don't nest exit tests
```

Author: This isn't an instruction. It's physically impossible to do it. There is no "do".

> An exit test cannot run within another exit test.

Reviewer: ```suggestion
An exit test can't run another exit test.
```

It means the same to say "a test can't run a child test" as "a test can't run in a parent test", and it's simpler.
